cronworker/cronworker.go

```diff
@@ -0,0 +1,207 @@
+package cronworker
+
+import (
+	"encoding/json"
+	"fmt"
+	"taas/dbstore"
+	"taas/diagnostics"
+	"taas/structs"
+	"time"
+
+	"github.com/lib/pq"
+	"github.com/robfig/cron/v3"
+	uuid "github.com/satori/go.uuid"
+)
+
+/*
+	The cron worker is only responsible for scheduling and running cron diagnostics.
+	It does not update configuration or add new cron jobs to the database.
+
+	It listens to the diagnosticdb using the Postgres NOTIFY/LISTEN pattern.
+	When records change in the cronjob table, it updates the internal cron scheduler.
+*/
+
+// Incoming message from the Postgres database
+// Indicates a database record has changed
+type incomingMessage struct {
+	Table  string `json:"table"`
+	Action string `json:"action"`
+	Data   struct {
+		OldRecord structs.Cronjob `json:"old_record"`
+		NewRecord structs.Cronjob `json:"new_record"`
+	} `json:"data"`
+}
+
+var cronScheduler *cron.Cron
+var jobmap map[string]cron.EntryID
+var listener *pq.Listener
+
+// Start starts the cron scheduler and listens for changed database records
+func Start() {
+	fmt.Println("[cron_worker]: Initializing...")
+
+	// Initialize the cron scheduler
+	cronScheduler = cron.New()
+	cronScheduler.Start()
+
+	dbstore.InitCronjobPool()
+
+	jobmap = make(map[string]cron.EntryID)
+
+	// Remove any old scheduled runs in database
+	err := dbstore.DeleteAllCronScheduleEntries()
+	if err != nil {
+		fmt.Println(err)
+	}
+
+	// Get a list of current database entries and add to cron scheduler
+	fmt.Println("[cron_worker]: Adding existing cronjobs from the database (if any)...")
+	var cronjobs []structs.Cronjob
+	cronjobs, err = dbstore.GetCronjobs()
+	if err != nil {
+		fmt.Println(err)
+	}
+
+	for _, element := range cronjobs {
+		if element.Disabled {
+			continue
+		}
+		err := addCronjob(element)
+		if err != nil {
+			fmt.Println(err)
+		}
+	}
+
+	// Create database listener
+	listener = dbstore.CreateListener()
+	err = listener.Listen("events")
+	if err != nil {
+		panic(err)
+	}
+
+	fmt.Println("[cron_worker]: Initialization complete!")
+	fmt.Println("[cron_worker]: Watching for database changes...")
+
+	// Wait and handle new messages from the database
+	for {
+		waitForChanges()
+	}
+}
+
+// Handle incoming database messages
+func waitForChanges() {
+	for {
+		select {
+
+		// Message received
+		case n := <-listener.Notify:
```

Comment:
What is `n` exactly? Is there a more descriptive variable name we can give this variable?

Reply:
it's a `pq.Notification`, changed var name